### PR TITLE
CORE-9517: Retry registration if P2P didn't work

### PR DESCRIPTION
**Pull request by yift-r3**
Retry the registration request a few times if there is no reply from the MGM.

It was tested by adding a random exit (i.e. `return`) in the MGM `StartRegistrationHandler`. When doing that, the registration was sent again, and passed successfully. It was also tested by using a wrong URL (port 8081 instead of 8080) and waiting for the request to fail.

It will try all the registration request types (not just members in a dynamic network). Still, because the static registration and MGM dynamic registration should set the state as part of the registration, the condition (i.e. if the state is still `SENT_TO_MGM`) will fail and not do the actual registration.

API pull request in https://github.com/corda/corda-api/pull/920

### CODE REVIEW

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/CommandsRetryManager.kt [hunk 1]
```diff
@@ -89,22 +87,22 @@ internal class CommandsRetryManager(
...
+            state.retryAfter.toEpochMilli() -
+                clock.instant().toEpochMilli()
         )
         val event = Record(
             MEMBERSHIP_ASYNC_REQUEST_TOPIC,
             key,
             state.request,
         )
-        logger.info("Request $key will be retried in $durationInMillis milliseconds")
+        logger.debug("Request $key will be retried in $durationInMillis milliseconds")
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/CommandsRetryManager.kt]: Should we keep this as `INFO` level?

I am thinking replays should be quite infrequent (e.g. only under failures / connectivity issues / misconfigured environments) and it would be useful to have that in the logs when investigating.

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt [hunk 2]
```diff
@@ -144,9 +150,18 @@ internal class StartRegistrationHandler(
                 key = "${mgmMemberInfo.holdingIdentity.shortHash}-${pendingMemberInfo.holdingIdentity.shortHash}",
                 value = persistentMemberInfo,
             )
+            val persistMemberStatusMessage = p2pRecordsFactory.createAuthenticatedMessageRecord(
```
Comment (dimosr) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt:176]: I think we can do without the extra message here. The MGM will send the verification request right after this and the member will switch their status to `PENDING_MEMBER_VERIFICATION` as part of processing it.

So, the replay logic on the member side can be conditional based on that, i.e. if the status is at that or a later step.

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt [hunk 3]
```diff
@@ -80,7 +96,11 @@ internal class MemberOpsAsyncProcessor(
     override val eventValueClass = MembershipAsyncRequest::class.java
     override val stateValueClass = MembershipAsyncRequestState::class.java
 
-    private fun register(request: RegistrationAsyncRequest, canRetry: Boolean): Outcome {
+    private fun register(
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: At a high-level, I think there are a few things that could be improved:
* we currently have logic based on the number of attempts for both cases. I think what we should be doing instead is 
  * having a number of attempts for the case where the request has failed and is retried for that reason, similar to what we had before.
  * for the case where the request is retried because it hasn't been confirmed as received from the MGM yet, replay periodically (every ~40 seconds as you have it now would be reasonable) with a timeout instead of counting the number of attempts. This timeout should be based on the TTL set by the MGM when sending the verification message, it should be higher than that to ensure the MGM has given up by that point. And if we do that, I think we should remove the `verifyMemberRequest` configuration option for membership to ensure both sides use compatible timeout configurations.

I think the schemas for the state will need some restructuring to do the above as it will probably be useful to represent explicitly these two categories (instead of having a single one with a retry condition).

**Review by dimosr — COMMENTED** [review]
Left a few high-level comments. One more question I have is what happens when/if registration requests from a replay arrive in the MGM after having already been processed? Are they handled properly at the moment or do we need to do some work there too (e.g. if StartRegistrationHandler logic is not fully idempotent atm, a simple fix could be to check if the registration already exists in the DB or Kafka at the `RegistrationRequestHandler` and skip sending the command)?

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/CommandsRetryManager.kt (hunk 1 above)
Comment (yift-r3) [reply components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/CommandsRetryManager.kt]: It's very noisy because we are getting this for every request. I changed it to be less noisy but I think we get a message when there is an actual replay.

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt (hunk 2 above)
Comment (yift-r3) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt:176]: We can, but I want to differentiate between a network issue and any other potential issue that might cause the `PENDING_MEMBER_VERIFICATION` not to be sent.

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt (hunk 3 above)
Comment (yift-r3) [reply components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: I've added the timeout explicitly to the configuration.

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt [hunk 4]
```diff
@@ -3,139 +3,225 @@ package net.corda.membership.service.impl
...
+            return if (retries <= 0) {
+                logger.warn(
+                    "Registration ${request.requestId} failed." +
+                        " Could not find holding identity associated with ${request.holdingIdentityId} for too long." +
+                        " Will not retry nor persist the state.",
+                )
+                createFailureWithoutRetryResponse()
+            } else {
+                logger.warn(
+                    "Registration ${request.requestId} failed." +
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: `"Registration ${request.requestId} failed."` -> that is a bit misleading given we'll retry the request later on.
I'd suggest just saying `"Could not find holding identity associated with ${request.holdingIdentityId}. Request will be retried later."`

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt [hunk 5]
```diff
@@ -3,139 +3,225 @@ package net.corda.membership.service.impl
...
+        val holdingIdentityShortHash = ShortHash.of(request.holdingIdentityId)
+        val holdingIdentity =
+            virtualNodeInfoReadService.getByHoldingIdentityShortHash(holdingIdentityShortHash)?.holdingIdentity
+        if (holdingIdentity == null) {
+            val retries = state.retries()
+            return if (retries <= 0) {
+                logger.warn(
+                    "Registration ${request.requestId} failed." +
+                        " Could not find holding identity associated with ${request.holdingIdentityId} for too long." +
+                        " Will not retry nor persist the state.",
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]:
```suggestion
                    "Registration ${request.requestId} failed." +
                        " Could not find holding identity associated with ${request.holdingIdentityId} for too long." +
                        " Request will not be retried further.",
```
to make it a bit easier for the person that reads the logs and is unaware of internal details.

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt [hunk 6]
```diff
@@ -3,139 +3,225 @@ package net.corda.membership.service.impl
...
+    private fun persistAndCreateFailureWithoutRetryResponse(
+        holdingIdentity: HoldingIdentity,
+        registrationId: UUID,
+        message: String?,
+    ): StateAndEventProcessor.Response<MembershipAsyncRequestState> {
+        // CORE-10367: return the status update command as part of the onNext
+        membershipPersistenceClient.setRegistrationRequestStatus(
+            holdingIdentity,
+            registrationId.toString(),
+            RegistrationStatus.INVALID,
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: I think it might be worth introducing an extra type here (e.g. `FAILED`) to differentiate from cases where the request itself is invalid, e.g. because of incorrect data.

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt [hunk 7]
```diff
@@ -3,139 +3,225 @@ package net.corda.membership.service.impl
...
+                    "Registration ${request.requestId} failed." +
+                        " Could not find holding identity associated with ${request.holdingIdentityId}",
+                )
+                createFailureWithRetryResponse(value, retries)
+            }
+        }
+        val cause = state?.cause
+        if (cause is SentToMgmWaitingForNetwork) {
+            if (cause.stopRetriesAfter.isBefore(clock.instant())) {
+                val message = "Registration request ${state.request.request} had been sent to the MGM too long"
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]:
```suggestion
                val message = "Registration request ${state.request.request} was not received by the MGM after many attempts."
```

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt [hunk 8]
```diff
@@ -3,139 +3,225 @@ package net.corda.membership.service.impl
...
+            updatedState = null,
+            responseEvents = emptyList(),
+            markForDLQ = true,
+        )
+    private fun persistAndCreateFailureWithoutRetryResponse(
+        holdingIdentity: HoldingIdentity,
+        registrationId: UUID,
+        message: String?,
+    ): StateAndEventProcessor.Response<MembershipAsyncRequestState> {
+        // CORE-10367: return the status update command as part of the onNext
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: need to remember to address these assuming CORE-10367 is merged first.

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt [hunk 9]
```diff
@@ -3,139 +3,225 @@ package net.corda.membership.service.impl
...
-                request.requestId
+                request.requestId,
             ).getOrThrow()
-            if ((requestStatus != null) && (requestStatus.registrationStatus != RegistrationStatus.NEW)) {
-                // This request had already passed this state. no need to continue.
-                return Outcome.SUCCESS
+            if (cause is SentToMgmWaitingForNetwork) {
+                if ((requestStatus == null) ||
+                    (requestStatus.registrationStatus == RegistrationStatus.NEW) ||
+                    (requestStatus.registrationStatus == RegistrationStatus.SENT_TO_MGM)
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: should we just check against `SENT_TO_MGM` here given the dynamic member registration converts it to that synchronously (and the state will have that value only if registration has been processed successfully once)?

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt [hunk 10]
```diff
@@ -3,139 +3,225 @@ package net.corda.membership.service.impl
...
+            responseEvents = emptyList(),
+            markForDLQ = false,
+        )
+    }
+
+    private fun MembershipAsyncRequestState?.retries(): Int {
+        return (this?.cause as? RetriableFailure)?.numberOfRetriesToGo ?: MAX_RETRIES
+    }
+
+    private fun register(
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: There are a lot of conditionals dispersed in this method now that makes it quite hard to follow, so worth thinking how it can be cleaned up a bit.

Some suggestions from me are 1. bringing all the conditionals related to checking if the request shouldn't be retried at the top (e.g. in case of retryable failure, crossed threshold of retries and in case of sent to mgm, time out) and extracting them into a separate method. This might require a small adjustment of the logic to exit early when retries are exhausted, instead of checking when there's a failure (I think that was the way you had it before) 2. a similar thing could be done for checks on whether the registration request has already moved on that follow on, which could be grouped and extracted in a separate method.

**Review by dimosr — COMMENTED** [review]
Worth adding some more details on the results of this testing:
> It was tested by adding a random exit in the MGM StartRegistrationHandler.

For example, where was the exit added and what was the result of the replays on the MGM side.

One more scenario worth testing (which is one that can happen in practice when that's merged) is what happens when a member registers with an invalid endpoint. In that case, the member will re-send the registration request several times before giving up, so we need to validate the MGM will be able to process them gracefully.

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt (hunk 10 above)
Comment (yift-r3) [reply components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: I broke it into a few methods. Hopefully, it will be cleaner.

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt (hunk 9 above)
Comment (yift-r3) [reply components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: It's asynchronous now.

Comment (dimosr) [reply components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: ah right, it was changed as part of https://github.com/corda/corda-runtime-os/pull/3289.

is the intention to check here we're not past `SENT_TO_MGM`? If so, would it be slightly cleaner if we do that by comparing the order of the current status to check that it's not after the order of `RECEIVED_BY_MGM`?

Comment (dimosr) [reply components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: And we could do something similar for the check below too, i.e. check order of current status is after order of `NEW`?

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt (hunk 10 above)
Comment (dimosr) [reply components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: yeah much better now, ordering the methods in the order they are invoked / their level of abstraction would help with reading the file a bit more (e.g. `register` is at the bottom after all the lower-level methods now). But, that's just a minor thing.

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt [hunk 11]
```diff
@@ -214,16 +214,15 @@ internal class MemberOpsAsyncProcessor(
         ).getOrThrow()
         if (cause is SentToMgmWaitingForNetwork) {
             if ((requestStatus == null) ||
-                (requestStatus.registrationStatus == RegistrationStatus.NEW) ||
-                (requestStatus.registrationStatus == RegistrationStatus.SENT_TO_MGM)
+                (requestStatus.registrationStatus <= RegistrationStatus.SENT_TO_MGM)
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]: is that comparing the order of the declaration of the enums in their definition? I think it'd be better if we introduce a method in `RegistrationStatusHelper` that compares based on the semantic order defined there or expose the `getOrder` method and use that to do the comparison. 

It doesn't make a practical difference here as for these statuses both orders are the same, but it will matter if we follow the same practice elsewhere for phases that might have the same semantic order, but different order in the definition (e.g. `INVALID` and `APPROVED`).

**Comment by dimosr** [comment]
There seems to be one (minor) [unaddressed comment](https://github.com/corda/corda-runtime-os/pull/3311#discussion_r1158319315). Other than that, it looks fine to me.

One question about testing with invalid endpoint though: what was the status of the registration request on the MGM side, i.e. was it stuck to `PENDING_MEMBER_VERIFICATION`? It wasn't clear to me from the description. Not a big problem for this PR, but if that's the case should we have a follow-up ticket to switch these registrations to a terminal status (e.g. `FAILED`) on the MGM side after a timeout expires?

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt [hunk 12]
```diff
@@ -3,136 +3,279 @@ package net.corda.membership.service.impl
...
+                        " Request will not be retried further.",
+                )
+                createFailureWithoutRetryResponse(
+                    emptyList(),
+                )
+            } else {
+                logger.warn(
+                    "Registration ${request.requestId} failed." +
+                        " Could not find holding identity associated with ${request.holdingIdentityId}",
+                )
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsAsyncProcessor.kt]:
```suggestion
                logger.warn(
                    "Could not find holding identity associated with ${request.holdingIdentityId}. " +
                            "Request ${request.requestId} will be retried later."
                )
```

**Review by dimosr — COMMENTED** [review]
I think the last change was applied to the wrong place. Added a fresh comment to indicate what the initial suggestion was again.

**Comment by yift-r3** [comment]
If the registration request from the member to the MGM is lost, the MGM is unaware of the request and has no status.
If the reply from the MGM to the member gets lost, the status will end up as `DECLINED`.

**Comment by dimosr** [comment]
There seems to be an error in `MultiClusterDynamicNetworkTest`, where a registration gets declined. Needs some investigation to understand what is going on. 

However, from a quick look at the logs it seems that the registration is replayed on the member side and potentially handled multiple times on the MGM side. This made me think about [this comment](https://github.com/corda/corda-runtime-os/pull/3311#pullrequestreview-1332777969) again; should we be adding an early check in the `RegistrationRequestHandler` to check if the registration request has already been persisted (i.e. from the `StartRegistrationHandler` when the message arrived previously) and in these cases just discard it considering it's a duplicate and must have been processed already? Otherwise, I am afraid not all of the checks are idempotent at the moment and it could lead to registrations being declined if processed again.

**Comment by yift-r3** [comment]
I think the decline's root cause is the link manager's duplication (https://github.com/corda/corda-runtime-os/pull/3596). I will add a safeguard in this pull request.

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt [hunk 13]
```diff
@@ -77,6 +77,10 @@ internal class StartRegistrationHandler(
     override val commandType = StartRegistration::class.java
 
     override fun invoke(state: RegistrationState?, key: String, command: StartRegistration): RegistrationHandlerResult {
+        if (state != null) {
```
Comment (dimosr) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt:80]: @nikinagy FYI, when we get to merge the re-registration branch, this deduplication check will probably need to be moved around, since state seems to be created earlier then. Maybe inside the `QueueRegistrationHandler`. 

Most likely it will come up as a conflict, but I thought it would be worth letting you know anyway.

**Review by dimosr — DISMISSED** [review]
LGTM - might be worth triggering a few runs of the tests before merging to ensure it doesn't introduce any flakiness.

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt (hunk 13 above)
Comment (nikinagy) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt:80]: @dimosr noted, thanks for letting me know.

**Review by vkolomeyko — APPROVED** [review]

**Review by dimosr — APPROVED** [review]
